collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java

```diff
@@ -0,0 +1,432 @@
+                    }
+                    if (downloadUrl == null || formName == null) {
+                        String error =
+                                "Forms list entry " + Integer.toString(i)
+                                        + " is missing form name or url attribute";
+                        Timber.e("Parsing OpenRosa reply -- %s", error);
+                        formList.clear();
+                        formList.put(
+                                DL_ERROR_MSG,
+                                new FormDetails(Collect.getInstance().getString(
+                                        R.string.parse_legacy_formlist_failed, error)));
+                        return formList;
+                    }
+                    formList.put(formName,
+                            new FormDetails(formName, downloadUrl, null, formId, null, null, null, false, false));
+
+                    formId = null;
+                }
+            }
+        }
+        return formList;
+    }
+
+    private static boolean isThisFormAlreadyDownloaded(String formId) {
+        return new FormsDao().getFormsCursorForFormId(formId).getCount() > 0;
+    }
+
+    private static ManifestFile getManifestFile(String manifestUrl) {
+        if (manifestUrl == null) {
+            return null;
+        }
+
+        // get shared HttpContext so that authentication and cookies are retained.
+        HttpContext localContext = Collect.getInstance().getHttpContext();
+
+        HttpClient httpclient = WebUtils.createHttpClient(WebUtils.CONNECTION_TIMEOUT);
+
+        DocumentFetchResult result =
+                WebUtils.getXmlDocument(manifestUrl, localContext, httpclient);
+
+        if (result.errorMessage != null) {
+            return null;
+        }
+
+        String errMessage = Collect.getInstance().getString(R.string.access_error, manifestUrl);
+
+        if (!result.isOpenRosaResponse) {
+            errMessage += Collect.getInstance().getString(R.string.manifest_server_error);
+            Timber.e(errMessage);
+            return null;
+        }
+
+        // Attempt OpenRosa 1.0 parsing
+        Element manifestElement = result.doc.getRootElement();
+        if (!manifestElement.getName().equals("manifest")) {
+            errMessage +=
+                    Collect.getInstance().getString(R.string.root_element_error,
+                            manifestElement.getName());
+            Timber.e(errMessage);
+            return null;
+        }
+        String namespace = manifestElement.getNamespace();
+        if (!FormDownloader.isXformsManifestNamespacedElement(manifestElement)) {
+            errMessage += Collect.getInstance().getString(R.string.root_namespace_error, namespace);
+            Timber.e(errMessage);
+            return null;
+        }
+        int elements = manifestElement.getChildCount();
+        List<MediaFile> files = new ArrayList<>();
+        for (int i = 0; i < elements; ++i) {
```

Comment:
This loop could be extracted out to another method, helping to shorten getManifestFile.

Reply:
Maybe but it's the code I just moved from one class to another. Since this pr is pretty big itself we should do such a refactoring in a separate one.